Lombiq.Hosting.Azure.ApplicationInsights/Extensions/ApplicationInsightsInitializerExtensions.cs

```diff
@@ -38,16 +38,11 @@ public static class ApplicationInsightsInitializerExtensions
 
         services.Configure<TelemetryConfiguration>(config =>
         {
-            if (applicationInsightsOptions.EnableLocalDevelopment)
+            if (!applicationInsightsOptions.EnableLocalDevelopment)
```

Comment:
We also need authentication during local development. "Local authentication" (somewhat confusingly named) means that using an Entra ID identity is not required (though allowed) and you can use API keys and connection strings without Entra ID. It doesn't mean that the control channel will be authenticated without anything else during local development.

You can now check out the [Local OSOCE AI resource](https://portal.azure.com/#@lombiq.com/resource/subscriptions/f04789ac-5112-4bb1-8725-eb44038c5273/resourcegroups/OpenSourceOrchardCoreExtensions-EUS/providers/microsoft.insights/components/OpenSourceOrchardCoreExtensions-EUS-Local/quickPulse). If you see this, then the control channel is not authenticated:


https://github.com/Lombiq/Orchard-Azure-Application-Insights/assets/1976647/6c90abf6-19fb-4f05-8ea4-e4d087310968

Note that securing this control channel, i.e. the AI account in the Azure Portal telling the app what telemetry to send, is the point of this issue. Telemetry, including live metrics, can be sent from the app to AI without this too (as seen happening here as well).

Reply:
I have looked into this more today. 

If you disable the confusingly named `LOCAL AUTHENTICATION` the only way to authenticate the local client to the AI resource is setting up a Service Principal (One of 2 options described [here](https://learn.microsoft.com/en-us/azure/azure-monitor/app/azure-ad-authentication?tabs=net#configure-and-enable-microsoft-entra-id-based-authentication))

The documentation in the readme and naming I have slightly changes to hopefully make things more clear.

The way I understand it is that it comes down to basically 4 scenario's now:

1. If you have `Local Authentication` enabled, all you need to do is supply the connectionstring (Not Authenticated)
2. If you have `Local Authentication` disabled, you need to set up a service principal and supply the details in order to securely push metrics from a locally hosted service.
3. For a service hosted on Azure (AppService, VM or other) you need to set up a Managed Identity, supply it with the Metrics Publisher Role and with that the `DefaultAzureCredential` will work to publish metrics securely. 
4. For now also supporting QuickPulseTelemetryModuleAuthenticationApiKey which is the deprecated method.

These are the scenario's I have tested today, please let me know if I have missed anything.

Reply:
In reply to: https://github.com/Lombiq/Orchard-Azure-Application-Insights/pull/85#discussion_r1628492342
Receiving Live Metrics with just connection string does still work as described in scenario 1, however this does no longer work if you disable `Local Authentication` I have tested and confirmed this locally.

![image](https://github.com/Lombiq/Orchard-Azure-Application-Insights/assets/1669059/3d7062bc-6eeb-4e6c-8840-ec688d9b6229)
As soon as you disable this you can no longer publish live metrics by just using the connectionstring, you are then forced to then use either a service principal, managed identity or API key (deprecated). 
Also see [here](https://learn.microsoft.com/en-us/azure/azure-monitor/app/azure-ad-authentication?WT.mc_id=Portal-AppInsightsExtension&tabs=net#disable-local-authentication)

> Both b) and c) satisfy Local Authentication being disabled. Right? Or is b) still Local Authentication?

As for Service Principal and Managed identity these will both work with Local Authentication being disabled. Only service principal can work locally though because you can't have a managed identity locally. 

So bottom line is that with `Local Authentication` disabled like shown above. The only way to publish live metrics from a locally running application (not in azure) in the future will be using the service principal.

With `Local Authentication` enabled, Using just the Connection String still works. (This is how it currently is for the OpenSourceOrchardCoreExtensions-EUS-Local resource)

In reply to: https://github.com/Lombiq/Orchard-Azure-Application-Insights/pull/85#discussion_r1628485291
I agree this is a little confusing, partially also because of Microsofts naming. And perhaps I will have to rethink this a little. Service Principal is the only way to publish live metrics once `Local Authentication` is disabled. However Service Principal is also technically a valid way to do it inside of Azure (although Managed Identity is recommended and easier to set up in this case). So I guess this should be renamed to UseServicePrincipal or something instead.

Reply:
> Receiving Live Metrics with just connection string does still work as described in scenario 1.

Just the connection string doesn't work, I tried before my last comment, no Live Metrics. I guess because if you don't set up SP credentials, then `DefaultAzureCredential` will be used, which won't work locally and breaks it.

> Only service principal can work locally because you can't have a managed identity locally.

Yes, but that's not what I'm asking. My question is, if Local Authentication is disabled, will an SP work, or only Managed Identity? So, together with your last paragraph please clear up in the docs and config names the following scenarios: What should I configure in the app, and what should I configure in the Azure Portal (can I disable Local Authentication)?

1.  I want to only ingest data in AI and don't care about the control channel. This is most local development scenarios but can very well be for deployed apps too. I think this scenario is the same for any hosting environment (you just have to provide the connection string) but please check.
2.  I want to both ingest data and secure the control channel, in an environment where Managed Identity is NOT available (locally or Azure Services not supporting it).
3.  I want to both ingest data and secure the control channel, in an environment where Managed Identity is available.

Reply:
I made some changes today to ensure that using just connectionstring still works without any changes.
I did have to introduce another option into the appsettings though because I couldn't see any other way to support just using connectionstring without it.

I have also made more edits to the readme to hopefully make it more clear what needs to be set up in which scenario.

To recap:
1) This will work with just connectionstring as long as `Local Authentication` is enabled. I've also tried to make this more clear in the readme
2) This requires a SP, again I've tried to make this more clear in the readme now.
3) This now requires `"UseEntraAuthentication": true` in the appsettings, the readme is updated to reflect this.

If the readme is still not clear enough maybe we can just add these 3 scenario's as a sort of FAQ style section at the bottom of the readme something like "Frequent scenario's" and the exact steps to take?

Reply:
Great, then! I don't think a FAQ, specifically, is needed, but please emphasize which option is necessary/available in which scenario (understanding that people will come with a scenario in mind, a use case, not an understanding of what Local Authentication/Service Principals/Managed Identity is).

Also add validation to throw an exception with an invalid combination of settings (like you can't have SP and MI at the same time).

Reply:
> Also add validation to throw an exception with an invalid combination of settings (like you can't have SP and MI at the same time).

I am not entirely sure how I would go about this. If you enabled `UseEntraAuthentication` but have not enabled `UseServicePrincipalAuthentication` it will default to try `DefaultAzureCredential` which will fail silently if it can't resolve any and you just will not see any metrics.

However if you have `UseEntraAuthentication`  and `UseServicePrincipalAuthentication` it will try the SP credentials this will already throw an error if it's not a valid configuration.

The only extra check I can do is to see if you have 'UseServicePrincipalAuthentication' enabled without having `UseEntraAuthentication`  enabled - which would be an invalid configuration.

Reply:
Perhaps UseEntraAuthentication should then be more like UseManagedIdentityAuthentication? And it's either-or, and both are Entra. Though at this point rather having an enum for the two and a single property would be better.